pkg/cmd/cli/cmd/volume.go

```diff
@@ -0,0 +1,585 @@
+		NamespaceParam(v.DefaultNamespace).DefaultNamespace().
+		SelectorParam(v.Selector).
+		ResourceTypeOrNameArgs(v.All, args...).
+		Flatten()
+
+	one := false
+	infos, err := b.Do().IntoSingular(&one).Infos()
+	if err != nil {
+		return err
+	}
+
+	skipped := 0
+	for _, info := range infos {
+		ok, err := v.UpdatePodSpecForObject(info.Object, func(spec *kapi.PodSpec) error {
+			var e error
+			switch {
+			case v.Add:
+				e = v.addVolumeToSpec(spec)
+			case v.Remove:
+				e = v.removeVolumeFromSpec(spec)
+			case v.List:
+				e = v.listVolumeForSpec(spec, info)
+			}
+			return e
+		})
+		if !ok {
+			skipped++
+			continue
+		}
+		if err != nil {
+			fmt.Fprintf(v.Writer, "error: %s/%s %v\n", info.Mapping.Resource, info.Name, err)
+			continue
+		}
+	}
+	if one && skipped == len(infos) {
+		return fmt.Errorf("the %s %s is not a pod or does not have a pod template", infos[0].Mapping.Resource, infos[0].Name)
+	}
+
+	if v.List {
+		return nil
+	}
+
+	objects, err := resource.AsVersionedObject(infos, false, v.OutputVersion)
+	if err != nil {
+		return err
+	}
+
+	if len(v.Output) != 0 {
+		p, _, err := kubectl.GetPrinter(v.Output, "")
+		if err != nil {
+			return err
+		}
+		return p.PrintObj(objects, v.Writer)
+	}
+
+	failed := false
+	for _, info := range infos {
+		data, err := info.Mapping.Codec.Encode(info.Object)
+		if err != nil {
+			fmt.Fprintf(v.Writer, "Error: %v\n", err)
+			failed = true
+			continue
+		}
+		obj, err := resource.NewHelper(info.Client, info.Mapping).Update(info.Namespace, info.Name, true, data)
+		if err != nil {
+			v.handleUpdateError(err)
+			failed = true
+			continue
+		}
+		info.Refresh(obj, true)
+		fmt.Fprintf(v.Writer, "%s/%s\n", info.Mapping.Resource, info.Name)
+	}
+	if failed {
+		return errExit
+	}
+	return nil
+}
+
+func (v *VolumeOptions) setVolumeSource(kv *kapi.Volume) error {
+	opts := v.AddOpts
+	if len(opts.Type) > 0 {
+		switch strings.ToLower(opts.Type) {
+		case "emptydir":
+			kv.EmptyDir = &kapi.EmptyDirVolumeSource{}
+		case "hostpath":
+			kv.HostPath = &kapi.HostPathVolumeSource{
+				Path: opts.Path,
+			}
+		case "secret":
+			kv.Secret = &kapi.SecretVolumeSource{
+				SecretName: opts.SecretName,
+			}
+		case "persistentvolumeclaim":
+			kv.PersistentVolumeClaimVolumeSource = &kapi.PersistentVolumeClaimVolumeSource{
+				ClaimName: opts.ClaimName,
+			}
+		default:
+			return fmt.Errorf("invalid volume type: %s", opts.Type)
+		}
+	} else if len(opts.Source) > 0 {
+		err := json.Unmarshal([]byte(opts.Source), &kv.VolumeSource)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (v *VolumeOptions) addVolumeToSpec(spec *kapi.PodSpec) error {
+	opts := v.AddOpts
+	newVolume := &kapi.Volume{
+		Name: v.Name,
+	}
+	setSource := true
+	for i, vol := range spec.Volumes {
+		if v.Name == vol.Name {
+			if !opts.Overwrite {
+				return fmt.Errorf("Volume '%s' already exists. Use --overwrite to replace", v.Name)
+			}
+			if !typeChanged && len(opts.Source) == 0 {
```

Comment:
If we don't remove and re-add the volume in case of overwrite, we need to unset the existing volume and I don't want to take that route (reason: we may not be able to check all available volume sources assuming there will be addition of new volume sources in the future).